# Week1_3_DataManagement_3

For this lab, we will learn: 
   - Concatenating multiple Dataframes along the rows
   - Merging dataframes along the columns

# 1 Concatenating multiple Dataframes along the rows

Pandas provides various functions for combining together Series or DataFrame along the rows or columns axis. 
   - "concatenate along the rows" means joining DataFrame B to DataFrame A vertically using column names as concatenating/joining identifiers.
   - Let us try to concatenate the 2012 and 2013 Beijing housing dataset along the rows using `pd.concat()`. 
    
    
To concatenate two DataFrames along the rows, you are requried to check whether the **joining identifiers (column names) are the same in both datasets**. 

Now read the two **cleaned** datasets we want to concatenate and print out their column names:  

## 1.1 Preparing to cancatenate two dataframes

In [3]:
# always remember to import the required libraries first
import pandas as pd

In [1]:
import os
path = os.getcwd()

In [13]:
# read two cleaned housing datasets
df_2012 = pd.read_excel(path + "/HouseBeijing2012_clean.xlsx")
df_2013 = pd.read_excel(path + "/HouseBeijing2013_clean.xlsx")

In [14]:
# print out columns names
print(df_2012.columns)
print(df_2013.columns)

Index(['Unnamed: 0', 'HouseID', 'CommunityID', 'TotalPrice', 'TransYear',
       'Bedroom', 'Livingroom', 'Bathroom', 'Size', 'FloorLevel', 'WinSouth',
       'WinSouthNorth', 'Decoration', 'TotalFloor', 'BuiltYear', 'Elevation',
       'Heating', 'TransMonth', 'TransDay', 'District', 'CensusTract',
       'XIAOQUWEB', 'SchQuality', 'NumSubway1km', 'Dist2Subway', 'HospQuality',
       'Dist2Hosp', 'NumHosp1km', 'NumBus200m', 'Dist2CBD', 'Dist2Center',
       'UnitPrice'],
      dtype='object')
Index(['1. HouseID', '2. CommunityID', '3. TotalPrice', '4. TransYear',
       '5. Bedroom', '6. Livingroom', '7. Bathroom', '8. Size',
       '9. FloorLevel', '10. WinSouth', '11. WinSouthNorth', '12. Decoration',
       '13. TotalFloor', '14. BuiltYear', '15. Elevation', '16. Heating',
       '17. TransMonth', '18. TransDay', '20. District', '21. CensusTract',
       '22. XIAOQUWEB', '23. SchQuality', '24. NumSubway1km',
       '25. Dist2Subway', '26. HospQuality', '27. Dist2Hosp', '28. NumHosp

In [18]:
print(len(df_2012.columns),len(df_2013.columns))

32 30


In [15]:
# if it is hard to identify if the columns are same, we can use: 
list(df_2012.columns) == list(df_2013.columns)

False

Two datasets are different: 
   - two more columns in the df_2012 dataset
   - the existence of ordinal number in the column names of 2013 dataset. 
   

## 1.2 Concatenate two dataframes

To concatenate two datasets, we need to make sure the column names are the same.
   - Step  1: drop two extra columns - `'Unnamed: 0'` `'UnitPrice'`
   - Step 2: assign the column names of the 2012 DataFrame to that of 2013 DataFrame
   - Step 3: apply `.concat()` to combine the two.  

In [20]:
# Step 1:
df_2012.drop(columns=['Unnamed: 0', 'UnitPrice'], inplace=True)

- `inplace = True` modifies the DataFrame in place, meaning it will directly apply the changes to the original DataFrame without needing to assign it to a new variable.

In [21]:
# Step 2: 
# get the column name of 2012 dataframe
col_2012 = df_2012.columns 

# assign the column names of 2012 dataset to the column names of 2013 dataset
df_2013.columns = col_2012 

In [22]:
# Step 3: Concate (Join) 2012 and 2013 dataset along the rows using pd.concat() 
df_comb_1213 = pd.concat([df_2012, df_2013]) 
df_comb_1213

HouseID  CommunityID  TotalPrice  TransYear  Bedroom  Livingroom  \
0     BJFT84326414          1544  1400010.56       2012        2           1   
1     BJDX84905788          2264  1350038.34       2012        2           1   
2     BJFT00386624          3621  1800006.91       2012        2           1   
3     BJCY84713854          1127  1970019.58       2012        1           1   
4     BJCY84112518          2767  2730093.60       2012        3           2   
...             ...          ...         ...        ...      ...         ...   
4995  BJCY85295028          3506  3400028.52       2013        2           2   
4996  BJFT86470175          2974  2040030.00       2013        2           1   
4997  BJCY85539965          1680  1970005.96       2013        1           1   
4998  BJDX85632173          3293  3150106.60       2013        3           2   
4999  BJTJ86774022          1714  2120057.94       2013        2           1   

      Bathroom    Size  FloorLevel  WinSouth  ...  \
0            1   69.68           1         0  ...   
1            1   88.83           3         1  ...   
2            1   98.69           3         0  ...   
3            1   53.66           4         0  ...   
4            2  132.08           2         1  ...   
...        ...     ...         ...       ...  ...   
4995         2  111.48           2         0  ...   
4996         1   57.00           5         0  ...   
4997         1   43.94           4         1  ...   
4998         2  130.45           1         1  ...   
4999         1   94.38           3         1  ...   

                                         XIAOQUWEB  SchQuality  NumSubway1km  \
0     https://bj.lianjia.com/xiaoqu/1111027377493/           0             2   
1     https://bj.lianjia.com/xiaoqu/1111027379274/           0             1   
2     https://bj.lianjia.com/xiaoqu/1111027382765/           0             1   
3     https://bj.lianjia.com/xiaoqu/1111027376538/           0             3   
4     https://bj.lianjia.com/xiaoqu/1111027380518/           0             1   
...                                            ...         ...           ...   
4995  https://bj.lianjia.com/xiaoqu/1111027382459/           0             2   
4996  https://bj.lianjia.com/xiaoqu/1111027381044/           0             0   
4997  https://bj.lianjia.com/xiaoqu/1111027377814/           0             1   
4998  https://bj.lianjia.com/xiaoqu/1111027381865/           0             0   
4999  https://bj.lianjia.com/xiaoqu/1111027377893/           0             0   

      Dist2Subway  HospQuality    Dist2Hosp  NumHosp1km  NumBus200m  \
0       633.24007            9   1803.02071           0           1   
1       667.21572            8  11158.05983           0           4   
2       939.29061            9   1698.79101           0          10   
3       476.28267            9    938.35742           2           0   
4       692.53062            9   5122.56097           0           0   
...           ...          ...          ...         ...         ...   
4995    718.61206            9   3072.19586           0           0   
4996   1096.16421            9   1841.03871           0           0   
4997    577.23111            7    292.66932           1           7   
4998   1183.40088            8   2576.24539           0           5   
4999   3936.33494            9   1667.86016           0           2   

         Dist2CBD  Dist2Center  
0      9345.20091   7396.31505  
1     22480.82065  20105.06770  
2     16309.85203  11427.48851  
3      8105.90581   7213.87518  
4     11524.68492  16967.87510  
...           ...          ...  
4995   7698.13101   8724.78408  
4996   7833.44901   4972.72498  
4997   5901.86525   6733.63486  
4998  11878.33320  12565.01538  
4999  18100.89223  23595.76940  

[9997 rows x 30 columns]

Argument: `axis`
- **axis = 0 means concating along rows, meaning joining two dataframes vertically This is the default.** 
- **axis = 1 means concating along columns**

In [23]:
# try axis = 0
df_comb_1213 = pd.concat([df_2012, df_2013], axis = 0) 
df_comb_1213

HouseID  CommunityID  TotalPrice  TransYear  Bedroom  Livingroom  \
0     BJFT84326414          1544  1400010.56       2012        2           1   
1     BJDX84905788          2264  1350038.34       2012        2           1   
2     BJFT00386624          3621  1800006.91       2012        2           1   
3     BJCY84713854          1127  1970019.58       2012        1           1   
4     BJCY84112518          2767  2730093.60       2012        3           2   
...             ...          ...         ...        ...      ...         ...   
4995  BJCY85295028          3506  3400028.52       2013        2           2   
4996  BJFT86470175          2974  2040030.00       2013        2           1   
4997  BJCY85539965          1680  1970005.96       2013        1           1   
4998  BJDX85632173          3293  3150106.60       2013        3           2   
4999  BJTJ86774022          1714  2120057.94       2013        2           1   

      Bathroom    Size  FloorLevel  WinSouth  ...  \
0            1   69.68           1         0  ...   
1            1   88.83           3         1  ...   
2            1   98.69           3         0  ...   
3            1   53.66           4         0  ...   
4            2  132.08           2         1  ...   
...        ...     ...         ...       ...  ...   
4995         2  111.48           2         0  ...   
4996         1   57.00           5         0  ...   
4997         1   43.94           4         1  ...   
4998         2  130.45           1         1  ...   
4999         1   94.38           3         1  ...   

                                         XIAOQUWEB  SchQuality  NumSubway1km  \
0     https://bj.lianjia.com/xiaoqu/1111027377493/           0             2   
1     https://bj.lianjia.com/xiaoqu/1111027379274/           0             1   
2     https://bj.lianjia.com/xiaoqu/1111027382765/           0             1   
3     https://bj.lianjia.com/xiaoqu/1111027376538/           0             3   
4     https://bj.lianjia.com/xiaoqu/1111027380518/           0             1   
...                                            ...         ...           ...   
4995  https://bj.lianjia.com/xiaoqu/1111027382459/           0             2   
4996  https://bj.lianjia.com/xiaoqu/1111027381044/           0             0   
4997  https://bj.lianjia.com/xiaoqu/1111027377814/           0             1   
4998  https://bj.lianjia.com/xiaoqu/1111027381865/           0             0   
4999  https://bj.lianjia.com/xiaoqu/1111027377893/           0             0   

      Dist2Subway  HospQuality    Dist2Hosp  NumHosp1km  NumBus200m  \
0       633.24007            9   1803.02071           0           1   
1       667.21572            8  11158.05983           0           4   
2       939.29061            9   1698.79101           0          10   
3       476.28267            9    938.35742           2           0   
4       692.53062            9   5122.56097           0           0   
...           ...          ...          ...         ...         ...   
4995    718.61206            9   3072.19586           0           0   
4996   1096.16421            9   1841.03871           0           0   
4997    577.23111            7    292.66932           1           7   
4998   1183.40088            8   2576.24539           0           5   
4999   3936.33494            9   1667.86016           0           2   

         Dist2CBD  Dist2Center  
0      9345.20091   7396.31505  
1     22480.82065  20105.06770  
2     16309.85203  11427.48851  
3      8105.90581   7213.87518  
4     11524.68492  16967.87510  
...           ...          ...  
4995   7698.13101   8724.78408  
4996   7833.44901   4972.72498  
4997   5901.86525   6733.63486  
4998  11878.33320  12565.01538  
4999  18100.89223  23595.76940  

[9997 rows x 30 columns]

The combined dataset looks good! It has 9,997 rows by 30 columns. But each row still has the original index that ranges from 0 to 4999. We need to reset the index. 

reindex the rows using the function `df.reset_index(drop = False, inplace = False)`. 
   - The arguments `drop = True` means dropping the original index column.
   - The argument `inplace = True` means that the dataframe is modified (reindexed) directly so that we do not need to assign a new variable to store the returned value. 
   - If `inplace = False` or using the default, the dataframe will not change after reindexing, and we have to assign a new variable to save the change

In [24]:
# reset the index
df_comb_1213.reset_index(drop = True, inplace = True) 
df_comb_1213

HouseID  CommunityID  TotalPrice  TransYear  Bedroom  Livingroom  \
0     BJFT84326414          1544  1400010.56       2012        2           1   
1     BJDX84905788          2264  1350038.34       2012        2           1   
2     BJFT00386624          3621  1800006.91       2012        2           1   
3     BJCY84713854          1127  1970019.58       2012        1           1   
4     BJCY84112518          2767  2730093.60       2012        3           2   
...             ...          ...         ...        ...      ...         ...   
9992  BJCY85295028          3506  3400028.52       2013        2           2   
9993  BJFT86470175          2974  2040030.00       2013        2           1   
9994  BJCY85539965          1680  1970005.96       2013        1           1   
9995  BJDX85632173          3293  3150106.60       2013        3           2   
9996  BJTJ86774022          1714  2120057.94       2013        2           1   

      Bathroom    Size  FloorLevel  WinSouth  ...  \
0            1   69.68           1         0  ...   
1            1   88.83           3         1  ...   
2            1   98.69           3         0  ...   
3            1   53.66           4         0  ...   
4            2  132.08           2         1  ...   
...        ...     ...         ...       ...  ...   
9992         2  111.48           2         0  ...   
9993         1   57.00           5         0  ...   
9994         1   43.94           4         1  ...   
9995         2  130.45           1         1  ...   
9996         1   94.38           3         1  ...   

                                         XIAOQUWEB  SchQuality  NumSubway1km  \
0     https://bj.lianjia.com/xiaoqu/1111027377493/           0             2   
1     https://bj.lianjia.com/xiaoqu/1111027379274/           0             1   
2     https://bj.lianjia.com/xiaoqu/1111027382765/           0             1   
3     https://bj.lianjia.com/xiaoqu/1111027376538/           0             3   
4     https://bj.lianjia.com/xiaoqu/1111027380518/           0             1   
...                                            ...         ...           ...   
9992  https://bj.lianjia.com/xiaoqu/1111027382459/           0             2   
9993  https://bj.lianjia.com/xiaoqu/1111027381044/           0             0   
9994  https://bj.lianjia.com/xiaoqu/1111027377814/           0             1   
9995  https://bj.lianjia.com/xiaoqu/1111027381865/           0             0   
9996  https://bj.lianjia.com/xiaoqu/1111027377893/           0             0   

      Dist2Subway  HospQuality    Dist2Hosp  NumHosp1km  NumBus200m  \
0       633.24007            9   1803.02071           0           1   
1       667.21572            8  11158.05983           0           4   
2       939.29061            9   1698.79101           0          10   
3       476.28267            9    938.35742           2           0   
4       692.53062            9   5122.56097           0           0   
...           ...          ...          ...         ...         ...   
9992    718.61206            9   3072.19586           0           0   
9993   1096.16421            9   1841.03871           0           0   
9994    577.23111            7    292.66932           1           7   
9995   1183.40088            8   2576.24539           0           5   
9996   3936.33494            9   1667.86016           0           2   

         Dist2CBD  Dist2Center  
0      9345.20091   7396.31505  
1     22480.82065  20105.06770  
2     16309.85203  11427.48851  
3      8105.90581   7213.87518  
4     11524.68492  16967.87510  
...           ...          ...  
9992   7698.13101   8724.78408  
9993   7833.44901   4972.72498  
9994   5901.86525   6733.63486  
9995  11878.33320  12565.01538  
9996  18100.89223  23595.76940  

[9997 rows x 30 columns]

## 1.3 Combining multiple dataframes

The following code:
   - uses `for loop`, `if statement` and `pd.concat()` to combine the five cleaned housing datasets from 2012 to 2016;
   - export the combined DataFrame to an Excel file.
   
How to design the for loop?
   - We first observe whether the names of these five Excel files have any common or different parts. We see that except for the **year**, all other parts are the same. This is great! In the for loop, we can iterate over the **year** to read each file. 
   - we use `+ str(year) +` to construct the relative file path:
       - `str(year)` means transforming the `year = 2012, ..., 2016` from an integer to a string such that `year = "2012",...,"2016"`. 
       - The `+` signs at the both ends link two parts of the path. As such, each integral year can be converted to string and forms the relative file path, for example, `"housing_dataset/HouseBeijing2012.xlsx"` 
       
Why use `if statement`?
   - We need to make sure the column names are the same for all  

In [25]:
# create a blank dataframe
df_combine = pd.DataFrame() 

# loop over 2012 - 2016
for year in list(range(2012, 2017)): 
    
    # construct the relative file path
    df_temp = pd.read_excel("HouseBeijing_clean" + str(year) + ".xlsx")
    
    # if statement
    if year == 2012:
        
        df_temp.drop(columns=['Unnamed: 0', 'UnitPrice'], inplace=True)
        
        # assign df_2012 to df_combine
        df_combine = df_temp 
        
        # save the column names
        col_2012 = df_temp.columns
        
    else:    
        
        # assign col_2012 to the column names of 2012 dataset to the column names of 2013 dataset
        df_temp.columns = col_2012
        
        # concatenate
        df_combine = pd.concat([df_combine, df_temp])

# reset the index
df_combine.reset_index(drop = True, inplace = True) 

# export the combined dataframe
df_combine.to_excel("housing_combine.xlsx")
df_combine

OSError: [Errno 22] Invalid argument: 'HouseBeijing_clean2012.xlsx'

# 2 Merging DataFrames along the columns

Now we want to merge DataFrames along the columns. we can use the function `pd.merge()`.
   - `merge along the columns` means merging two dataframes horizontally.
   - Note that the `pd.concat()` can be used to merge along columns by changing the argument `axis = 1`; however, my personal perference is `pd.merge()`.
   - Also note, the function `pd.merge()` can **ONLY** be used to merge along the columns.
   
In the `df_combine`, you may notice the column (variable) "District" is coded by numbers, 1,2,3... without acknowledging us the specific district names. There is a file called `"DistrictName.xlsx"` in on Canvas storing the district name. We will merge the DataFrame with distrcit name to the current DataFrame. So that we know where a housing record is located.

To merge the two DataFrames, we need to 
   - identify the merging key (identifier), which is the common column existing in both DataFrames.  
   - the merging key in the current DataFrame (df_combine) is the column "District" and the merging key in the district DataFrame is the column "DistrictID".  


Check the merging key in the `df_combine`: column "District".

In [ ]:
# examine the variable "District"
df_combine["District"]

Read the `DistrictName.xlsx` and examine the merging keys.

In [ ]:
# read the District name file. 
df_dist = pd.read_excel("/DistrictName.xlsx")
df_dist.head(5) 

# DistrictID as merge key

In [ ]:
# merge the combined dataframe with the district name
# left_on: the left merge key, right_on: the right merge key
df_combine = pd.merge(df_combine, df_dist, left_on = "District", right_on = "DistrictID")
df_combine

`pd.merge()` is a powerful tool for combining dataframe along the columns. We are unable to throughly explore all its features and we do not need to. 

   - pd.merge(self, right, how='inner', on=None, left_on=None, right_on=None, left_index=False, right_index=False, sort=False, suffixes=('_x', '_y'), copy=True, indicator=False, validate=None)
      - pandas `pd.merge()` help document: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html
 
***`If you have a concatenate/merge task with more requirements (e.g., merging based on multiple keys, or one to many, one to many, many to many, etc.), check tutorial here:` https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html. `It lists all the possible scenarios of combining DataFrames that can be done using pandas.`***

# 3 Slicing string columns

`Key Identifer` is important in merging and concentating multiple dataframes. However, finding the appropriate identifer is not easy. Sometimes, we need to slice the string column, keeping only parts of the string, in order to match with the identifer in another dataframe. 

For example, in our dataframe, we have a column "HouseID", storing the unqiue housing transcation number. A typical houseID, such as BJFT84326414, consists of two parts: the first part is a four-digit city/district code (e.g., BJFT), and the second part is a unique transaction number for each house. We now only need the transaction number and need to remove the city/district code, or place this part into another column.

In [5]:
# read the cleaned housing datasets
df_2012 = pd.read_excel(path + "/HouseBeijing2012_clean.xlsx")

In [7]:
df_2012[["HouseID"]]

HouseID
0     BJFT84326414 
1     BJDX84905788 
2     BJFT00386624 
3     BJCY84713854 
4     BJCY84112518 
...             ...
4992  BJTJ84718789 
4993  BJFT84287006 
4994  BJDC84781079 
4995  BJSJ85075781 
4996  BJCY84949599 

[4997 rows x 1 columns]

To do this, we can: 
- (1) create a new column saving the city/district code
- (2) create another column saving housing transaction number.
- (3) drop the original HouseID

In [9]:
# Extract city and district code (first 4 digits)
df_2012['HouseID'].str[:4]

0       BJFT
1       BJDX
2       BJFT
3       BJCY
4       BJCY
        ... 
4992    BJTJ
4993    BJFT
4994    BJDC
4995    BJSJ
4996    BJCY
Name: HouseID, Length: 4997, dtype: object

In [11]:
# Extract transaction number (remaining part)
df_2012['HouseID'].str[4:]

0       84326414 
1       84905788 
2       00386624 
3       84713854 
4       84112518 
          ...    
4992    84718789 
4993    84287006 
4994    84781079 
4995    85075781 
4996    84949599 
Name: HouseID, Length: 4997, dtype: object

In [ ]:
# Drop the original houseID column if needed
df_2012.drop(columns=['HouseID'], inplace=True)